src/linear_algebra/basis.lean

```diff
@@ -1424,3 +1424,60 @@ let ⟨q, hq⟩ := p.exists_is_compl in nonempty.intro $
   (prod_equiv_of_is_compl q p hq.symm)
 
 end division_ring
+
+section restrict_scalars
+
+variables [comm_ring R] [ring K] [nontrivial K] [add_comm_group V]
+variables [algebra R K] [module K V] [module R V]
+variables [is_scalar_tower R K V] [no_zero_smul_divisors R K] (b : basis ι K V)
+variable (R)
+
+open submodule
+
+/-- Let `b` be a `K`-basis of `V`. Let `R` be a comm_ring such that `algebra R K` with no zero
+smul divisors, then the submodule of `V` spanned by `b` over `R` admits `b` as a `R`-basis. -/
+noncomputable def basis.restrict_scalars : basis ι R (span R (set.range b)) :=
+basis.span (b.linear_independent.restrict_scalars (smul_left_injective R one_ne_zero))
+
+@[simp]
+lemma basis.restrict_scalars_apply (i : ι) : (b.restrict_scalars R i : V) = b i :=
+by simp only [basis.restrict_scalars, basis.span_apply]
+
+@[simp]
+lemma basis.restrict_scalars_repr_apply (m : span R (set.range b)) (i : ι) :
+  algebra_map R K ((b.restrict_scalars R).repr m i) = b.repr m i :=
```

Comment:
I wonder if this is easier to prove by going via
```
finsupp.map_range.linear_map (algebra.linear_map R K) ∘ₗ (b.restrict_scalars R).repr.to_linear_map
    = ((b.repr : V →ₗ[K] (ι →₀ K)).restrict_scalars R).dom_restrict _
```

Reply:
I see. I was looking indeed for this results :)

That should help a lot.